src/Abstractions/TaskName.cs

```diff
@@ -29,6 +29,26 @@ public TaskName(string name)
         }
     }
 
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TaskName"/> struct.
+    /// </summary>
+    /// <param name="name">The name of the task. Providing <c>null</c> will yield the default struct.</param>
+    /// <param name="version">The version of the task.</param>
+    public TaskName(string name, string version)
+    {
+        if (name is null)
+        {
+            // Force the default struct when null is passed in.
+            this.Name = null!;
+            this.Version = version!;
```

Comment:
Thank you, @jviau, I follow your above two suggestions, but there is warning:

```c#
        if (name is null && version is null)
        {
            // Force the default struct when null is passed in.
            this.Name = null!;
            this.Version = null!;
        }
        else
        {
            this.Name = name;    // 'name' may be null here.
            this.Version = version;
        }
```

Reply:
hmm yeah need to think through the behavior a bit: what to do when `name` is `null`, but `version` isn't? That is probably an error case. Maybe `ArgumentException` explaining `name` must not be `null` when `version` is non-null.

Reply:
There are four case about name or version is null or not:

1. name is null & version is null

    ```c#
    this.Name = null!;
    this.Version = null!;
    ```
2. name is null & version is not null

    ```c#
    throw ArgumentException("name must not be null when version is non-null")
    ```

3. name is not null and version is null

    ```c#
    this.Name = name;
    this.Version = null!;
    ```

4. name is not null and version is null

    ```c#
    this.Name = name;
    this.Version =version;
    ```

Please confirm if above are right.

Reply:
Almost correct - just some nuance around `Version`. It is not null, so we should default to `string.Empty` (as we do currently).

- If version is null and name is not null, throw argument exception as well. (Users should pass in string.Empty as default).

Reply:
Thank, I had updated the code, please help to review. 

For the case that version is null and name is not null, I suggest not throw argument exception but fallback to the constructor without version parameter. 

I think it would be more friendly. Of cause, it depends on how we co "version is null and name is not null",  This shouldn't be considered an error. 

Of course, it depends on how we look at the case "version is null and name is not null", whether we think of version as an required parameter and should not be null, or whether we understand version as a non-required parameter that null value is acceptable and we will fallback to the constructor without version parameter.

Reply:
@jviau @cgillum please help to review it and let us make a decision, this PR is almost ready to merge except this.